fl4health/clients/apfl_client.py

```diff
@@ -1,255 +1,117 @@
-
-    def is_start_of_local_training(self, epoch: int, step: int) -> bool:
-        return epoch == 0 and step == 0
-
-    def fit(self, parameters: NDArrays, config: Config) -> Tuple[NDArrays, int, Dict[str, Scalar]]:
-        if not self.initialized:
-            self.setup_client(config)
-
-        self.set_parameters(parameters, config)
-        local_epochs = self.narrow_config_type(config, "local_epochs", int)
-
-        # Default APFL uses an average meter
-        global_meter = MetricAverageMeter(self.metrics, "global")
-        local_meter = MetricAverageMeter(self.metrics, "local")
-        personal_meter = MetricAverageMeter(self.metrics, "personal")
-        # By default the APFL client trains by epochs
-        metric_values = self.train_by_epochs(local_epochs, global_meter, local_meter, personal_meter)
-        # FitRes should contain local parameters, number of examples on client, and a dictionary holding metrics
-        # calculation results.
-        return (
-            self.get_parameters(config),
-            self.num_examples["train_set"],
-            metric_values,
-        )
-
-    def evaluate(self, parameters: NDArrays, config: Config) -> Tuple[float, int, Dict[str, Scalar]]:
-        if not self.initialized:
-            self.setup_client(config)
-
-        self.set_parameters(parameters, config)
-        # Default APFL uses an average meter
-        global_meter = MetricAverageMeter(self.metrics, "global")
-        local_meter = MetricAverageMeter(self.metrics, "local")
-        personal_meter = MetricAverageMeter(self.metrics, "personal")
-        loss, metric_values = self.validate(global_meter, local_meter, personal_meter)
-        # EvaluateRes should return the loss, number of examples on client, and a dictionary holding metrics
-        # calculation results.
-        return (
-            loss,
-            self.num_examples["validation_set"],
-            metric_values,
-        )
-
-    def _handle_logging(
-        self, loss_dict: Dict[str, float], metrics_dict: Dict[str, Scalar], is_validation: bool = False
-    ) -> None:
-        loss_string = "\t".join([f"{key}: {str(val)}" for key, val in loss_dict.items()])
-        metric_string = "\t".join([f"{key}: {str(val)}" for key, val in metrics_dict.items()])
-        metric_prefix = "Validation" if is_validation else "Training"
-        log(
-            INFO,
-            f"alpha: {self.model.alpha} \n"
-            f"Client {metric_prefix} Losses: {loss_string} \n"
-            f"Client {metric_prefix} Metrics: {metric_string}",
-        )
-
-    def train_step(self, input: torch.Tensor, target: torch.Tensor) -> ApflTrainStepOutputs:
+
+        # local_optimizer is used on the local model
+        # Usual self.optimizer is used for global model
+        self.local_optimizer: Optimizer
+
+    def is_start_of_local_training(self, step: int) -> bool:
+        return step == 0
+
+    def update_after_step(self, step: int) -> None:
```

Comment:
Do you mind filling in a bit of a comment here? Maybe note that the `step` passed here is assumed to represent the overall number of local training steps taken by the client in the current server round?

Reply:
Added/Updated some more comments in [e15e2f8](https://github.com/VectorInstitute/FL4Health/pull/62/commits/e15e2f8a3dac6f637c956cdbabf3265a1af6bd41)